sycl/source/detail/graph_impl.cpp

```diff
@@ -71,6 +73,40 @@ bool checkForRequirement(sycl::detail::AccessorImplHost *Req,
   }
   return SuccessorAddedDep;
 }
+
+/// Visits a node on the graph and it's successors recursively in a depth-first
+/// approach.
+/// @param[in] Node The current node being visited.
+/// @param[in,out] VisitedNodes A set of unique nodes which have already been
+/// visited.
+/// @param[in] NodeStack Stack of nodes which are currently being visited on the
+/// current path through the graph.
+/// @param[in] NodeFunc The function object to be run on each node. A return
+/// value of true indicates the search should be ended immediately and the
+/// function will return.
+/// @return True if the search should end immediately, false if not.
+bool visitNodeDepthFirst(
+    std::shared_ptr<node_impl> Node,
+    std::set<std::shared_ptr<node_impl>> &VisitedNodes,
+    std::deque<std::shared_ptr<node_impl>> &NodeStack,
+    std::function<bool(std::shared_ptr<node_impl> &,
+                       std::deque<std::shared_ptr<node_impl>> &)>
+        NodeFunc) {
+  auto EarlyReturn = NodeFunc(Node, NodeStack);
+  if (EarlyReturn) {
+    return true;
+  }
+  NodeStack.push_back(Node);
+  Node->MVisited = true;
```

Comment:
Where are we changing behaviour based on `MVisited`? I see we initalize it to zero, set it to `true` here, then reset it at the end of `graph_impl::searchDepthFirst`. Wondering if we can remove this variable

Reply:
This is more of a generic thing and is a fairly fundamental component of doing a DFS. It was used originally in the cycle check but that actually required the stack for the current path to function correctly. Conceivably it could be useful for other queries that use a DFS. 

I'd be in favour of keeping it for those reasons. It might also make sense to add an optional post-search (but before the visited status is reset) function which could potentially be useful or more efficient for queries that need it, but I'd probably leave that for if/when we actually need it.

Reply:
That makes sense, but would like a comment noting that it's current unused and with that rationale for its existence.